### PR TITLE
Feature/micro service support

### PR DESCRIPTION
**Pull request by Qblack**
Basically I wanted the ability to not need scipy or the UI stuff so I could have a lightweight package to put on microservices. 

I added a setup_micro.py that only takes what I think we need - as well as flush_cache on teardown. 
The requirements are also truncated.

### CODE REVIEW

**Comment by mircealungu** [comment]
Hi @Qblack ! 
Nice to hear that you find the FMD useful :)

And thanks for the PR!
We'll try to look at it and its sisters ASAP!

**Comment by Qblack** [comment]
Sounds good - I imagine this one will have some back and forth - the other 2 are much more straightforward.

File: requirements-micro.txt
Comment (mircealungu) [inline requirements-micro.txt:10]: if you look at the requirements.txt you'll see that it has changed and is now frozen to 2.0.

File: requirements-micro.txt
Comment (mircealungu) [inline requirements-micro.txt:17]: remove these two commented lines maybe?

File: setup_micro.py
Comment (mircealungu) [inline setup_micro.py:23]: Is it necessary to add a new project name in PyPI?

Comment (mircealungu) [reply setup_micro.py:23]: Beyond having a new project name, there's a lot of duplication between files now.
setup_micro.py and requirements_micro have to always be changed with the corresponding files, and that is not cool. Can you think of a better solution?

File: flask_monitoringdashboard/__init__.py
Comment (mircealungu) [inline flask_monitoringdashboard/__init__.py:81]: why do we need a teardown only on micro? 
shouldn't this be in there always?

**Review by mircealungu — COMMENTED** [review]
@Qblack - I finally had the chance to look a bit at the PR. I really like the idea of deploying for micro-service architectures. How do you exactly do this microservices setup? It would be cool if you could add a FMD-Micro.md maybe in which you describe this? I think it might be good for others too!

I've left a few comments on the code. Overall I like the idea of allowing for a setup_micro that does not deploy scipy (I guess the gain there is speed of deployment?). But I don't like the code duplication between setup*.py and requirements*.txt files. Do you think we can fix that somehow?

File: flask_monitoringdashboard/__init__.py
Comment (Qblack) [reply flask_monitoringdashboard/__init__.py:81]: So this is here to help with https://github.com/flask-dashboard/Flask-MonitoringDashboard/issues/276. 
The reason it was happening for me was because I am using one dashboard for multiple apps and the cache flush only gets triggered on the overview call of the api. This meant the last call time was always out of date unless I went to that apps dashboard. 
I have it under micro because that is when the problem is most prevelent and assumed someone would want to test it properly in a non micro context.

File: setup_micro.py
Comment (Qblack) [reply setup_micro.py:23]: RE: Adding new pypi project - Probaly? I have my own that I use and it needed a seperate wheel for sure. I have no idea if there is a way to do versions of your project. I think i saw someone do weird stuff with []s before though so could be a possible avenue to persue.

For the other point you are right. We could in theory import setup.py into setup micro and then modify it appropriatly in setup_micro - I assume we can pass an exploded dict to .setup. Alternatively we could have a common file that gets imported and modified by both. 

Requirements wise - probably the same deal - have a common reqs file and then a micro and standard might work?

I kinda took the path of least resistance for this since I wanted to use it myself right away and assumed you would have some feedback.

File: requirements-micro.txt
Comment (Qblack) [reply requirements-micro.txt]: Right. I think that happend after I started working on this. Added.

File: requirements-micro.txt
Comment (Qblack) [reply requirements-micro.txt]: Done

**Comment by FlyingBird95** [comment]
Hey @Qblack,

Thanks for showing interest in the FMD. I appreciate your enthousiasm in the project. 

About your PR, I have couple of general comments for you:

- First about the `requirements-micro.txt`: We indeed need two requirements files, but it is not ideal to list the dependencies in both of the files. As a solution, we can do the following:
  requirements-micro.txt:
  <list all requirements that are needed for the micro FMD>

  requirements.txt
  first line (this is the trick): add the following line `-r requirements-micro.txt`
  <list all requirements that the micro does not need>
- about the setup file. We should not have this as a separate file. There should be only one. I think that we can distinguish the install command by checking `sys.args`. I.e.:
  - to install the FMD normally: `python setup.py`.
  - to install the micro FMD: `python setup.py --micro`.

Can you address these comments?

I'm looking forward to your future PR's for the FMD.

Patrick

**Comment by Qblack** [comment]
@FlyingBird95 
Thanks Patrick - I definetly like youre idea more than mine and it seems to work on my end.

Q

File: setup.py
Comment (FlyingBird95) [inline setup.py:12]: you don't need to remove it from the args

File: setup.py
Comment (FlyingBird95) [inline setup.py:27]: this can be a one-liner:
`required.extend([l for l in f.read().splitlines() if l != '-r requirements-micro.txt'])`

File: setup.py
Comment (FlyingBird95) [inline setup.py:49]: we should also exclude the tests in non-micro mode

File: setup.py
Comment (Qblack) [reply setup.py:12]: image

I could not get it to work without removing it. I think the extra arg was getting passed into setup. Althought to be fair I also probably make the wheels differently than you guys. So maybe thats why.

File: setup.py
Comment (Qblack) [reply setup.py]: Done

File: flask_monitoringdashboard/__init__.py
Comment (FlyingBird95) [inline flask_monitoringdashboard/__init__.py:32]: maybe we should name it `without_dashboard=False`?

File: flask_monitoringdashboard/__init__.py
Comment (FlyingBird95) [inline flask_monitoringdashboard/__init__.py:38]: I think this docstring deserves an update:
- ` - default False - `: can be removed.
- `True if you only want the db stuff` doesn't sound really professional to me :)

File: setup.py
Comment (Qblack) [reply setup.py:46]: For whatever reason find_packages has been giving me pain lately. I had thought you were excluding tests via specifying the test suite later - but maybe not? 

Because of the explicit includes up above these are technically not needed and likely just got copied from wherever the last time i used exclude was. Are you able to double check that it currently includes them on your end? Something in the way I am building could be funky.

File: flask_monitoringdashboard/__init__.py
Comment (Qblack) [reply flask_monitoringdashboard/__init__.py]: I can do that - I tried a couple different names but figured I should have it match with everything else being called micro.

File: flask_monitoringdashboard/__init__.py
Comment (Qblack) [reply flask_monitoringdashboard/__init__.py]: Fair. removed.

**Comment by FlyingBird95** [comment]
Nice work,

It looks good, so you have my approval.

I'm still not entirely sure about the flag `without_dashboard=False`. Maybe we should do `include_dashboard=True`. But I'll leave this up to you.

Thank you for contributing to the FMD, and looking forward to seeing more PR's from you.

Patrick

**Review by FlyingBird95 — APPROVED** [review]

**Comment by Qblack** [comment]
@FlyingBird95 ooh I like that alot better that way the if does not need a not. I only went with something that defaulted to false to try and be as unintrusive as possible.

**Comment by mircealungu** [comment]
Now I see that this PR is still not merged. 
@Qblack @FlyingBird95: Is there anything else to be done here, or should we merge it in?

**Comment by Qblack** [comment]
> Now I see that this PR is still not merged.
> @Qblack @FlyingBird95: Is there anything else to be done here, or should we merge it in?

Go for it :D

**Pull request merged**
